Mathlib/Tactic/NormNum/Factors.lean

```diff
@@ -0,0 +1,146 @@
+/-
+Copyright (c) 2021 Mario Carneiro. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Mario Carneiro, Eric Wieser
+-/
+import Mathlib.Data.Nat.Factors
+import Mathlib.Tactic.NormNum.Prime
+
+/-!
+# `simproc` for `Nat.primeFactorsList`
+
+Note that since `norm_num` can only produce numerals, we can't use it here.
+-/
+
+open Nat
+
+namespace Mathlib.Meta.NormNum
+
+/-- A partial proof of `primeFactorsList`.
+Asserts that `l` is a sorted list of primes, lower bounded by a prime `p`, which multiplies to
+`n`. -/
+def FactorsHelper (n p : ℕ) (l : List ℕ) : Prop :=
+  p.Prime → List.Chain (· ≤ ·) p l ∧ (∀ a ∈ l, Nat.Prime a) ∧ List.prod l = n
+
+/-! The argument explicitness in this section is chosen to make only the numerals in the factors
+list appear in the proof term. -/
+
+theorem FactorsHelper.nil {a : ℕ} : FactorsHelper 1 a [] := fun _ =>
+  ⟨List.Chain.nil, by rintro _ ⟨⟩, List.prod_nil⟩
+
+theorem FactorsHelper.cons_of_le
+    {n m : ℕ} (a : ℕ) {b : ℕ} {l : List ℕ} (h₁ : IsNat (b * m) n) (h₂ : a ≤ b)
+    (h₃ : minFac b = b) (H : FactorsHelper m b l) : FactorsHelper n a (b :: l) := fun pa =>
+  have pb : b.Prime := Nat.prime_def_minFac.2 ⟨le_trans pa.two_le h₂, h₃⟩
+  let ⟨f₁, f₂, f₃⟩ := H pb
+  ⟨List.Chain.cons h₂ f₁,
+    fun _ h => (List.eq_or_mem_of_mem_cons h).elim (fun e => e.symm ▸ pb) (f₂ _),
+    by rw [List.prod_cons, f₃, h₁.out, cast_id]⟩
+
+theorem FactorsHelper.cons
+    {n m : ℕ} {a : ℕ} (b : ℕ) {l : List ℕ} (h₁ : IsNat (b * m) n) (h₂ : Nat.blt a b)
+    (h₃ : IsNat (minFac b) b) (H : FactorsHelper m b l) : FactorsHelper n a (b :: l) :=
+  H.cons_of_le _ h₁ (Nat.blt_eq.mp h₂).le h₃.out
+
+theorem FactorsHelper.singleton (n : ℕ) {a : ℕ} (h₁ : Nat.blt a n) (h₂ : IsNat (minFac n) n) :
+    FactorsHelper n a [n] :=
+  FactorsHelper.nil.cons _ ⟨mul_one _⟩ h₁ h₂
+
+theorem FactorsHelper.same {n m : ℕ} (a : ℕ) {l : List ℕ}
+    (h : IsNat (a * m) n) (H : FactorsHelper m a l) :
+    FactorsHelper n a (a :: l) := fun pa =>
+  H.cons_of_le _ h le_rfl (Nat.prime_def_minFac.1 pa).2 pa
+
+theorem FactorsHelper.same_singleton (a : ℕ) : FactorsHelper a a [a] :=
+  FactorsHelper.nil.same _ ⟨mul_one _⟩
+
+theorem FactorsHelper.primeFactorsList_eq {n : ℕ} {l : List ℕ} (H : FactorsHelper n 2 l) :
+    Nat.primeFactorsList n = l :=
+  let ⟨h₁, h₂, h₃⟩ := H Nat.prime_two
+  have := List.chain'_iff_pairwise.1 (@List.Chain'.tail _ _ (_ :: _) h₁)
+  (List.eq_of_perm_of_sorted
+    (Nat.primeFactorsList_unique h₃ h₂) this (Nat.primeFactorsList_sorted _)).symm
+
+open Lean Elab Tactic Qq
+
+/-- Given `n` and `a` natural numerals, returns `(l, ⊢ factorsHelper n a l)`. -/
+partial def evalPrimeFactorsListAux
+    {en n' : Q(ℕ)} {ea a' : Q(ℕ)} (hn : Q(IsNat $en $n')) (ha : Q(IsNat $ea $a')) :
+    MetaM ((l : Q(List ℕ)) × Q(FactorsHelper $en $ea $l)) := do
+  let n := n'.natLit!
+  let ⟨hn0⟩ ← if h : 0 < n then pure <| PLift.up h else
+    throwError m!"{n'} must be positive"
```

Comment:
Why do we give up on computing `primeFactorsList 0` here?

Reply:
This is the `aux` function, which expects the `0` and `1` cases to already have been dispatched, such that `a` can be chosen as a prime less than `n`.

I'll make it private to make this a little more clear.